### ISSUE TITLE
Issues with gl_PerVertex matching between stages

### ISSUE THREAD

**Issue by mbechard** [KhronosGroup/glslang#2578]
This issue is related to issue #2243 and issue #1726. It seems that the contents of gl_PerVertex is very inconsistency filled in glslang, and the extensions that change this block don't seem to be consistently following the specs. From everything I've read, the interface matching rule from the OpenGL spec, 7.4.1 requires the gl_PerVertex structure to match between stages except for gl_PointSize. Various extensions break this rule (GL_NVX_multiview_per_view_attributes, GL_ARB_shader_viewport_layer_array, GL_NV_stereo_view_rendering) without explicitly stating new exceptions.
But, even if they had exceptions, in glslang many of these variables aren't included in gl_PerVertex consistently. gl_ViewportIndex and gl_Layer aren't included in gl_PerVertex at all, even though the extension states they should be. In other cases such as gl_SecondaryPositionNV, it's not included in the Vertex shader's output gl_PerVertex block, but is included in the Geom/Tess/Control shader's input gl_PerVertex block. It is then left outside the output gl_PerVertex blocks for the Geom/Tess/Control shader.

Now, with the more robust block checking that is happening with the recent vulkan_relaxed merge, the gl_PerVertex interface is going to fail to match a lot. Attached is a minimal example, if you run it with 
`glslangValidator -V --aml --amb cross.vert cross.geom`
You'll see it's failing due to the gl_PerVertex blocks not matching.

Should we just disable interface matching validations for gl_PerVertex?

[cross.zip](https://github.com/KhronosGroup/glslang/files/6152913/cross.zip)

**Comment by greg-lunarg** [comment]
It is certainly unacceptable to report correct user code as erroneous. That must be remedied immediately.

I would prefer a slightly more surgical remedy ie. one that only suppresses errors around inconsistent variables and extensions. In other words, I am ok if the missing or lost error messages involve inconsistent variables and extensions.

One thing I would like to avoid is losing valid error messages around consistent variables and extensions that we were getting before the vulkan_relaxed merge. If no quick remedy can avoid this, such regressions should be noted in an issue and addressed with urgency.

Any other missing error messages, including those involving inconsistent variables and extensions, should be noted in a second issue and can be addressed with less urgency.

Would this be possible?

**Comment by mbechard** [comment]
Ya for sure. There was a time during vulkan_relaxed where it started to be fine, due to commit 8111268, which blindly avoids some gl_PerVertex checking. I'll investigate why the errors came back more recently.
I'll see what the previous behavior was, and the checks pass if they involve mismatches on variables that are known to be mismatched in their declarations.

**Comment by mbechard** [comment]
PR #2581 does minimal checking based on cases I can make fail so far. In theory gl_ViewportMask, gl_SecondaryViewportMaskNV and gl_ViewportMaskPerViewNV could also error when tessellation shaders are in use, but so far I'm unable to make those errors appear since this codepath check doesn't get used due to the 'in gl_PerVertex' blocks not getting included in the aggregate returned by findLinkerObjects().

I think one of the reasons this bug didn't hit any of the existing test cases is due to how rarely multi-stage linking is tested in the test cases.

**Comment by mbechard** [comment]
To add to this bug, trying to do cross-stage linking with regular GL (adding a test case to Link.FromFile.cpp) fails since the gl_PerVertex is polluted with all kinds of compatibility values such as gl_FrontColor, even when using #version 460 core.

**Comment by greg-lunarg** [comment]
The original problem with the false errors on the shaders in cross.zip has been fixed with PR #2581.

This issue is now a catch-all for missing matching errors for certain elements of gl_PerVertex mentioned above whose inclusion in gl_PerVertex is not consistently handled, including but limited to gl_FrontColor, gl_ViewportMask, gl_ViewportIndex, gl_Layer, gl_SecondaryViewportMaskNV, gl_ViewportMaskPerViewNV, gl_SecondaryPositionNV, gl_PositionPerViewNV.

**Comment by krop** [comment]
Any news about this issue? glslang 11.4.0 breaks the gammaray build

**Comment by mbechard** [comment]
@krop, all the known failure cases were fixed. If you have another one I'll be happy to take a look and fix it.

**Comment by greg-lunarg** [comment]
Please supply the shader source when creating the issue as well as a glslangValidator command line that replicates the problem.

**Comment by krop** [comment]
https://github.com/KDAB/GammaRay/issues/635 contains the details

**Comment by greg-lunarg** [comment]
@mbechard I believe that all your fixes are in 11.4.0, so this would be a current issue.

**Comment by mbechard** [comment]
@greg-lunarg these shaders are specifically declared as `#version 150 core`, however the gl_PerVertex is getting polluted in the vertex shader with
`gl_ClipVertex 4-component vector of float ClipVertex gl_ClipVertex,  out 4-component vector of float FrontColor gl_FrontColor,  out 4-component vector of float BackColor gl_BackColor,  out 4-component vector of float FrontSecondaryColor gl_FrontSecondaryColor,  out 4-component vector of float BackSecondaryColor gl_BackSecondaryColor,  out 1-element array of 4-component vector of float TexCoord gl_TexCoord,  out float FogFragCoord gl_FogFragCoord`
It feels like the way forward it so clean up the declaration of gl_PerVertex, rather than adding workarounds/exceptions for these in the cross stage validation. What do you think

**Comment by greg-lunarg** [comment]
FYI, here is a command line to recreate the problem: glslangValidator -l widget_wireframe.vert widget_wireframe.geom

I will creating a new issue for this problem.

**Comment by mbechard** [comment]
I wonder if any other resources can be used to finish up this issue? A lot of the commits come from Nvidia or companies with more resources than my tiny company.

**Comment by greg-lunarg** [comment]
I don't think you are obligated to finish this. If you wish I can un-assign you. Then someone with more interest or resources might volunteer. Let me know.

**Comment by mbechard** [comment]
Yeah, that may be best for this one. I wouldn't want to mess with changes that Nvidia put in specifically for their extensions really. Our commit for vulkan_relaxed brought to light a lot of these issues since it added the cross-stage validation that didn't exist before. But I'm not sure I'm the best person to fix every one of those problems.
